File: landeployer-backend/app/services/ssh_service.py
"""
SSH服务
"""
import paramiko
from loguru import logger
from typing import Optional, Tuple
from app.config import settings

class SSHService:
    """SSH连接和命令执行服务"""
    
    @staticmethod
    def create_client(host_info: dict) -> paramiko.SSHClient:
        """创建SSH客户端"""
        client = paramiko.SSHClient()
        client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
        
        try:
            if host_info["auth_type"] == "password":
                client.connect(
                    hostname=host_info["ip"],
                    port=host_info["port"],
                    username=host_info["username"],
                    password=host_info["password"],
                    timeout=settings.SSH_TIMEOUT
                )
            elif host_info["auth_type"] == "privateKey":
                pkey = paramiko.RSAKey.from_private_key_file(
                    host_info["private_key"],
                    password=host_info.get("private_key_password")
                )
                client.connect(
                    hostname=host_info["ip"],
                    port=host_info["port"],
                    username=host_info["username"],
                    pkey=pkey,
                    timeout=settings.SSH_TIMEOUT
                )
            
            return client
        except Exception as e:
            logger.error(f"SSH连接失败: {e}")
            raise
    
    @staticmethod
    def execute_command(host_info: dict, command: str) -> Tuple[bool, str]:
        """
        执行SSH命令
        返回: (是否成功, 输出内容)
        """
        client = None
        try:
            client = SSHService.create_client(host_info)
            stdin, stdout, stderr = client.exec_command(command)
            
            exit_status = stdout.channel.recv_exit_status()
            output = stdout.read().decode('utf-8')
            error = stderr.read().decode('utf-8')
            
            if exit_status != 0:
                logger.warning(f"命令执行失败, 退出码: {exit_status}, 错误: {error}")
                return False, error if error else output
            
            return True, output
            
        except Exception as e:
            logger.error(f"执行命令失败: {e}")
            return False, str(e)
        finally:
            if client:
                client.close()
    
    @staticmethod
    def test_connection(host_info: dict) -> Tuple[bool, str, Optional[str], Optional[str]]:
        """
        测试连接
        返回: (是否成功, 消息, OS信息, Docker版本)
        """
        try:
            client = SSHService.create_client(host_info)
            
            # 获取系统信息
            stdin, stdout, stderr = client.exec_command("uname -a")
            os_info = stdout.read().decode('utf-8').strip()
            
            # 获取Docker版本
            stdin, stdout, stderr = client.exec_command("docker --version")
            docker_version = stdout.read().decode('utf-8').strip()
            
            client.close()
            
            return True, "连接成功", os_info, docker_version
            
        except Exception as e:
            logger.error(f"连接测试失败: {e}")
            return False, f"连接失败: {str(e)}", None, None
    
    @staticmethod
    def check_file_exists(host_info: dict, remote_path: str) -> bool:
        """检查文件是否存在"""
        command = f"test -f {remote_path} && echo 'exists' || echo 'not exists'"
        success, output = SSHService.execute_command(host_info, command)
        return success and output.strip() == 'exists'
    
    @staticmethod
    def check_dir_exists(host_info: dict, remote_path: str) -> bool:
        """检查目录是否存在"""
        command = f"test -d {remote_path} && echo 'exists' || echo 'not exists'"
        success, output = SSHService.execute_command(host_info, command)
        return success and output.strip() == 'exists'
    
    @staticmethod
    def upload_file(host_info: dict, local_path: str, remote_path: str) -> bool:
        """上传文件"""
        client = None
        try:
            client = SSHService.create_client(host_info)
            sftp = client.open_sftp()
            
            # 确保远程目录存在
            remote_dir = remote_path.rsplit('/', 1)[0]
            try:
                sftp.stat(remote_dir)
            except FileNotFoundError:
                # 创建目录
                SSHService.execute_command(host_info, f"mkdir -p {remote_dir}")
            
            # 上传文件
            sftp.put(local_path, remote_path)
            sftp.close()
            
            logger.info(f"文件上传成功: {local_path} -> {remote_path}")
            return True
            
        except Exception as e:
            logger.error(f"文件上传失败: {e}")
            return False
        finally:
            if client:
                client.close()

File: landeployer-server/src/main/java/com/landeployer/service/SSHService.java
package com.landeployer.service;

import com.jcraft.jsch.*;
import com.landeployer.entity.Host;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.io.*;
import java.util.Properties;

/**
 * SSH服务
 */
@Slf4j
@Service
public class SSHService {
    
    @Value("${app.ssh-timeout:30000}")
    private Integer sshTimeout;
    
    /**
     * 创建SSH连接
     */
    public Session createSession(Host host) throws JSchException {
        JSch jsch = new JSch();
        
        Session session = jsch.getSession(host.getUsername(), host.getIp(), host.getPort());
        
        // 认证方式
        if ("password".equals(host.getAuthType())) {
            session.setPassword(host.getPassword());
        } else if ("privateKey".equals(host.getAuthType())) {
            if (host.getPrivateKeyPassword() != null && !host.getPrivateKeyPassword().isEmpty()) {
                jsch.addIdentity("key", host.getPrivateKey().getBytes(), 
                        null, host.getPrivateKeyPassword().getBytes());
            } else {
                jsch.addIdentity("key", host.getPrivateKey().getBytes(), null, null);
            }
        }
        
        // 配置
        Properties config = new Properties();
        config.put("StrictHostKeyChecking", "no");
        session.setConfig(config);
        session.setTimeout(sshTimeout);
        
        return session;
    }
    
    /**
     * 执行命令
     */
    public String executeCommand(Host host, String command) throws Exception {
        Session session = null;
        ChannelExec channel = null;
        
        try {
            session = createSession(host);
            session.connect();
            
            channel = (ChannelExec) session.openChannel("exec");
            channel.setCommand(command);
            
            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
            ByteArrayOutputStream errorStream = new ByteArrayOutputStream();
            
            channel.setOutputStream(outputStream);
            channel.setErrStream(errorStream);
            
            channel.connect();
            
            // 等待执行完成
            while (!channel.isClosed()) {
                Thread.sleep(100);
            }
            
            String output = outputStream.toString("UTF-8");
            String error = errorStream.toString("UTF-8");
            
            int exitStatus = channel.getExitStatus();
            
            if (exitStatus != 0) {
                log.warn("命令执行失败，退出码: {}, 错误: {}", exitStatus, error);
                return error;
            }
            
            return output;
            
        } finally {
            if (channel != null) {
                channel.disconnect();
            }
            if (session != null) {
                session.disconnect();
            }
        }
    }
    
    /**
     * 测试连接
     */
    public boolean testConnection(Host host) {
        Session session = null;
        try {
            session = createSession(host);
            session.connect();
            return session.isConnected();
        } catch (Exception e) {
            log.error("SSH连接测试失败: {}", e.getMessage());
            return false;
        } finally {
            if (session != null) {
                session.disconnect();
            }
        }
    }
    
    /**
     * 上传文件
     */
    public void uploadFile(Host host, String localFile, String remotePath) throws Exception {
        Session session = null;
        ChannelSftp channel = null;
        
        try {
            session = createSession(host);
            session.connect();
            
            channel = (ChannelSftp) session.openChannel("sftp");
            channel.connect();
            
            // 确保远程目录存在
            String remoteDir = remotePath.substring(0, remotePath.lastIndexOf('/'));
            createRemoteDir(channel, remoteDir);
            
            // 上传文件
            channel.put(localFile, remotePath);
            
            log.info("文件上传成功: {} -> {}", localFile, remotePath);
            
        } finally {
            if (channel != null) {
                channel.disconnect();
            }
            if (session != null) {
                session.disconnect();
            }
        }
    }
    
    /**
     * 创建远程目录
     */
    private void createRemoteDir(ChannelSftp channel, String path) throws SftpException {
        String[] dirs = path.split("/");
        String currentPath = "";
        
        for (String dir : dirs) {
            if (dir.isEmpty()) continue;
            
            currentPath += "/" + dir;
            try {
                channel.cd(currentPath);
            } catch (SftpException e) {
                channel.mkdir(currentPath);
                channel.cd(currentPath);
            }
        }
    }
    
    /**
     * 检查文件是否存在
     */
    public boolean checkFileExists(Host host, String remotePath) throws Exception {
        String command = "test -f " + remotePath + " && echo 'exists' || echo 'not exists'";
        String result = executeCommand(host, command);
        return result != null && result.trim().equals("exists");
    }
    
    /**
     * 检查目录是否存在
     */
    public boolean checkDirExists(Host host, String remotePath) throws Exception {
        String command = "test -d " + remotePath + " && echo 'exists' || echo 'not exists'";
        String result = executeCommand(host, command);
        return result != null && result.trim().equals("exists");
    }
}

